_posts/2020-03-05-pulp-3-scale-testing.md

```diff
@@ -0,0 +1,232 @@
+* `avg time` - The average amount of time from the three runs.
+* `avg repos / second` - The average amount of RepositoryVersion's created per second.
+
+
+## Database and CPU Constrained
+
+The system's load falls mostly onto the DB and is CPU bound. All 16 workers of Pulp were engaged
+(as expected) and the postgresql processes auto-scales in terms of it's processes, but it only
+has so many CPU cores to serve 16 independent connections (one from each Pulp worker).
+
+Here's a screenshot showing the postgresql processes heavily loading the CPU above 80% and 90% along
+with the rq Pulp worker processes submitting the SQL to PostgreSQL. I interpret the rate-limiting
+component to be the CPUs the database has access to.
+
+![The ps -awfux output](../../../../images/pulp_ansible_scale_test/ps_output_showing_postgresql_and_rq.png)
+ 
+ 
+## Database Query Analysis
+
+I enabled [`auto_explain`](https://www.postgresql.org/docs/9.2/auto-explain.html) for postgresql
+with these settings in my postgresql.conf:
+
+```ini
+session_preload_libraries = 'auto_explain'
+auto_explain.log_min_duration = '1000ms'
+auto_explain.log_analyze = 'true'
+```
+
+During promotion workflow testing, it shows slow queries like this one:
+[https://explain.depesz.com/s/Wi2Q](https://explain.depesz.com/s/Wi2Q). Suggestions on how to
+improve this are welcome, but it's overall a large database to query against. One unexpected part of
+the query plan is [step #15](https://explain.depesz.com/s/Wi2Q#l15) which perform a sequential scan
+against a table with an Index. After some discussion in #postgresql, other users there pointed out
+that the postgresql query plan is optimized and in some cases a Seq Scan is faster depending on the
+number of expected rows to be matched, e.g. 4950 in this case. I confirmed the indexes were present
+also, and that a single query does use INDEX SCAN as expected.
+
+
+## Bigger Batches are Better
+
+All tasks have to flow through the resource manager to provide the locking guarantees Pulp workloads
+require. This has a non-trivial overhead, and you can see that as the batch sizes get bigger
+throughput increases a little bit.
+
+We should look at benchmarking the maximum task throughput of the resource manager in a separate
+blog post.
+
+
+## Horizontal Scaling would go Faster!
+
+This system used 16 Pulp workers and so the work was divided over 16 parallel processes. If the DB
+has enough resources to keep up and Pulp had more cores it could start more workers and they could
+also run in parallel.
+
+This means that those with more resources can scale up as much as their PostgreSQL cluster will.
+Also if you want to start a few workers Pulp will just work on the same workload over more time.
+
+An average RepositoryVersions per second is roughly 31.4 based on all the runs in the table. With
+16 workers, that is 31.4 / 16 =  1.96 ~ 2 repositories per second. To estimate the capability of
```

Comment:
This sentence is confusing. Probably want `created` after `RepositoryVersions`.

Reply:
ty, rewriteed in next push.